Flaky test: BindOperationTestCase.testSubtreeModifyUpdatesAuthInfo — Connection refused right after in-process server restart

### Summary

`BindOperationTestCase.testSubtreeModifyUpdatesAuthInfo` failed once on `build-maven (ubuntu-latest, 17)` with `java.net.ConnectException: Connection refused` on the very first `RemoteConnection` opened right after `TestCaseUtils.restartServer()`. All other 8 matrix jobs of the same run passed, and the failure is unrelated to the PR being tested (#754, BouncyCastle FIPS bump — the BC FIPS provider installed fine and the server started normally).

Failed run: https://github.com/OpenIdentityPlatform/OpenDJ/actions/runs/29724957232/job/88295808274

```
[ERROR] org.opends.server.core.BindOperationTestCase.testSubtreeModifyUpdatesAuthInfo -- Time elapsed: 5.325 s <<< FAILURE!
java.net.ConnectException: Connection refused
    at org.opends.server.tools.RemoteConnection.<init>(RemoteConnection.java:91)
    at org.opends.server.tools.RemoteConnection.<init>(RemoteConnection.java:84)
    at org.opends.server.core.BindOperationTestCase.testSubtreeModifyUpdatesAuthInfo(BindOperationTestCase.java:1582)
```

### Analysis

Timeline reconstructed from the test's server error log (`logs/errors`):

- 08:48:56 — the test starts and calls `TestCaseUtils.restartServer()`; the in-process restart takes ~5 s on the slow runner (08:48:56 → 08:49:01);
- 08:49:01 — `Started listening for new connections on LDAP Connection Handler 0.0.0.0 port 65535`;
- 08:49:01.4 — the test fails to connect to that port.

The connect and the listener bind land in the same second — a race between the restarted server and the first client connection. The wait-for-listen handshake in `LDAPConnectionHandler.start()` is not bulletproof: it uses a bare `waitListen.wait()` with no condition loop (an early return on a spurious wakeup/interrupt is possible), and the `!enabled` branch in `run()` notifies the waiter before any channel is bound. `RemoteConnection` connects exactly once, with no retry.

Notably, this same pair of tests was already de-flaked once in 3cc8bed3 (#727), but that fix addressed the POST_RESPONSE `AuthenticatedUsers` polling race, not the connect-after-restart race.

### Proposed fix

Either (or both):

- wrap the initial `RemoteConnection` creation in the existing `TestTimer.repeatUntilSuccess(...)` helper (same approach as #727) in both `testSubtreeModifyUpdatesAuthInfo` and `testSubtreeDeleteClearsAuthInfo` (same construct at `BindOperationTestCase.java:1513`);
- make `TestCaseUtils.restartServer()` poll the LDAP port until it actually accepts connections before returning.

--- a/opendj-server-legacy/src/main/java/org/opends/server/protocols/ldap/LDAPConnectionHandler.java
+++ b/opendj-server-legacy/src/main/java/org/opends/server/protocols/ldap/LDAPConnectionHandler.java
@@ -13,6 +13,7 @@
  *
  * Copyright 2006-2010 Sun Microsystems, Inc.
  * Portions Copyright 2011-2016 ForgeRock AS.
+ * Portions Copyrighted 2026 3A Systems, LLC.
  */
 package org.opends.server.protocols.ldap;
 
@@ -182,6 +183,14 @@ public final class LDAPConnectionHandler extends
    * returning.
    */
   private final Object waitListen = new Object();
+
+  /**
+   * Condition predicate for {@link #waitListen}: set once the selector thread
+   * has attempted to open the listen channels (successfully or not). Guarded
+   * by the {@link #waitListen} monitor; protects the start method against
+   * spurious wakeups.
+   */
+  private boolean listenAttempted;
 
   /** The friendly name of this connection handler. */
   private String friendlyName;
@@ -814,12 +823,16 @@ public final class LDAPConnectionHandler extends
 
       try
       {
-        waitListen.wait();
+        while (!listenAttempted)
+        {
+          waitListen.wait();
+        }
       }
       catch (InterruptedException e)
       {
         // If something interrupted the start its probably better
         // to return ASAP.
+        Thread.currentThread().interrupt();
       }
     }
   }
@@ -859,6 +872,7 @@ public final class LDAPConnectionHandler extends
           synchronized (waitListen)
           {
             starting = false;
+            listenAttempted = true;
             waitListen.notify();
           }
         }
@@ -882,6 +896,7 @@ public final class LDAPConnectionHandler extends
         // should be notified and resume its work in any cases.
         synchronized (waitListen)
         {
+          listenAttempted = true;
           waitListen.notify();
         }
 

--- a/opendj-server-legacy/src/test/java/org/opends/server/TestCaseUtils.java
+++ b/opendj-server-legacy/src/test/java/org/opends/server/TestCaseUtils.java
@@ -14,7 +14,7 @@
  * Copyright 2006-2010 Sun Microsystems, Inc.
  * Portions Copyright 2011-2016 ForgeRock AS.
  * Portions Copyright 2013 Manuel Gaupp
- * Portions Copyright 2018-2025 3A Systems, LLC
+ * Portions Copyright 2018-2026 3A Systems, LLC
  */
 package org.opends.server;
 
@@ -68,6 +68,7 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
 import java.util.logging.ConsoleHandler;
 import java.util.logging.Handler;
 import java.util.logging.LogManager;
@@ -119,6 +120,7 @@ import org.forgerock.opendj.server.embedded.EmbeddedDirectoryServer;
 import org.opends.server.util.BuildVersion;
 import org.opends.server.util.DynamicConstants;
 import org.opends.server.util.LDIFReader;
+import org.opends.server.util.TestTimer;
 
 import com.forgerock.opendj.util.OperatingSystem;
 
@@ -327,6 +329,7 @@ public final class TestCaseUtils {
       setupLoggers();
       writeBuildInfoFile();
       server.start();
+      waitForLdapListener();
       assertTrue(InvocationCounterPlugin.startupCalled());
       // Save config.ldif for when we restart the server
       backupServerConfigLdif();
@@ -607,6 +610,7 @@ public final class TestCaseUtils {
       restoreServerConfigLdif();
 
       server.start();
+      waitForLdapListener();
 
       clearJEBackends();
       initializeTestBackend(true);
@@ -623,6 +627,38 @@ public final class TestCaseUtils {
       e.printStackTrace(originalSystemErr);
       throw e;
     }
+  }
+
+  /**
+   * Waits until the LDAP connection handler actually accepts TCP connections.
+   * {@code LDAPConnectionHandler.start()} may return before the listen
+   * channel is bound (e.g. after an early wakeup of its wait-for-listen
+   * handshake), so the very first connection made right after a server
+   * (re)start could otherwise be refused.
+   */
+  private static void waitForLdapListener() throws Exception
+  {
+    new TestTimer.Builder()
+        .maxSleep(10, TimeUnit.SECONDS)
+        .sleepTimes(100, TimeUnit.MILLISECONDS)
+        .toTimer()
+        .repeatUntilSuccess(new TestTimer.CallableVoid()
+        {
+          @Override
+          public void call() throws Exception
+          {
+            try (Socket s = new Socket())
+            {
+              s.connect(new InetSocketAddress("127.0.0.1", getServerLdapPort()), 500);
+            }
+            catch (IOException e)
+            {
+              // repeatUntilSuccess() only retries on assertion errors.
+              throw new AssertionError("LDAP port " + getServerLdapPort()
+                  + " is not accepting connections yet", e);
+            }
+          }
+        });
   }
 
   /**

